pyop2/mpi.py

```diff
@@ -118,63 +135,272 @@ def delcomm_outer(comm, keyval, icomm):
+            print(f'| {null:39}| {0:5d} |')
         else:
-            refcount = comm_out.Get_attr(refcount_keyval)
+            refcount = comm.Get_attr(refcount_keyval)[0]
             if refcount is None:
-                raise ValueError("Inner comm without a refcount")
-            refcount[0] += 1
+                refcount = -999
+            print(f'| {comm.name:39}| {refcount:5d} |')
+
+
+class PyOP2Comm:
+    """ Suitable for using a PyOP2 internal communicator suitably
+    incrementing and decrementing the comm.
+    """
+    def __init__(self, comm):
+        self.comm = comm
+        self._comm = None
+
+    def __enter__(self):
+        self._comm = internal_comm(self.comm)
+        return self._comm
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        decref(self._comm)
+        self._comm = None
+
+
+def internal_comm(comm):
+    """ Creates an internal comm from the comm passed in
+    This happens on nearly every PyOP2 object so this avoids unnecessary
+    repetition.
+    :arg comm: A communicator or None
+
+    :returns pyop2_comm: A PyOP2 internal communicator
+    """
+    if comm is None:
+        # None will be the default when creating most objects
+        pyop2_comm = dup_comm(COMM_WORLD)
+    elif is_pyop2_comm(comm):
+        # Increase the reference count and return same comm if
+        # already an internal communicator
+        incref(comm)
+        pyop2_comm = comm
+    elif isinstance(comm, PETSc.Comm):
+        # Convert PETSc.Comm to mpi4py.MPI.Comm
+        pyop2_comm = dup_comm(comm.tompi4py())
+    elif comm == MPI.COMM_NULL:
+        # Ensure comm is not the NULL communicator
+        raise ValueError("MPI_COMM_NULL passed to internal_comm()")
+    elif not isinstance(comm, MPI.Comm):
+        # If it is not an MPI.Comm raise error
+        raise ValueError("Don't know how to dup a %r" % type(comm))
+    else:
+        pyop2_comm = dup_comm(comm)
+    return pyop2_comm
+
+
+def incref(comm):
+    """ Increment communicator reference count
+    """
+    assert is_pyop2_comm(comm)
+    refcount = comm.Get_attr(refcount_keyval)
+    refcount[0] += 1
+
+
+def decref(comm):
+    """ Decrement communicator reference count
+    """
+    if comm == MPI.COMM_NULL:
+        comm = FriendlyCommNull()
+    if not PYOP2_FINALIZED:
+        assert is_pyop2_comm(comm)
+        refcount = comm.Get_attr(refcount_keyval)
+        refcount[0] -= 1
+        if refcount[0] == 0 and not isinstance(comm, FriendlyCommNull):
+            dupped_comms.remove(comm)
+            free_comm(comm)
+    elif comm == MPI.COMM_NULL:
+        pass
+    else:
+        free_comm(comm)
+
+
+def dup_comm(comm_in):
+    """Given a communicator return a communicator for internal use.
+
+    :arg comm_in: Communicator to duplicate
+
+    :returns: An mpi4py communicator."""
+    assert not is_pyop2_comm(comm_in)
+
+    # Check if communicator has an embedded PyOP2 comm.
+    comm_out = comm_in.Get_attr(innercomm_keyval)
```

Comment:
Maybe `internal_comm` would be a more informative name.

Reply:
Fixed